src/solvers/refinement/string_refinement.cpp

```diff
@@ -1085,21 +1096,49 @@ static exprt negation_of_constraint(const string_constraintt &axiom)
   return negaxiom;
 }
 
+/// Result of the solver `supert` should not be interpreted literally for char
+/// arrays as not all indices are present in the index set.
+/// In the given expression, we populate arrays at the indices for which the
+/// solver has no constraint by copying values to the left.
+/// For example an expression `ARRAY_OF(0) WITH [1:=2] WITH [4:=3]` would
+/// be interpreted as `{ 2, 2, 3, 3, 3}`.
+/// \param expr: expression to interpret
+/// \param string_max_length: maximum size of arrays to consider
+/// \return the interpreted expression
+exprt concretize_arrays_in_expression(
```

Comment:
I've done that, it required implementing a new operation in expression iterator, because we want to skip to the next  sibling once we changed an expression as it does not make sense to visit children of the old expression.
We also added unit tests for this `next_sibling_or_parent` operation.